llvmPackages_15: Building with `clang++` using `libcxx`/`-stdlib=libc++` fails

### Describe the bug
Using `-lc++` with `clang++` seems to be broken with `llvmPackages_15.clang` and `llvmPackages_15.libcxxClang`. It fails with `ld: cannot find l: No such file or directory`. https://github.com/NixOS/nixpkgs/issues/213144 also notes this.

### Steps To Reproduce

Running `clang++ -stdlib=libc++ main.cpp` yields `-/nix/store/r2b9k28c6aghczpqfvh71y9xavm7rr68-binutils-2.39/bin/ld: cannot find l: No such file or directory`. The `main.cpp` contains just an empty `main()` and I am using this flake's devShell:
```nix
{
  description = "Test";

  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
  };

  outputs = { self, nixpkgs }: let
    lib = nixpkgs.lib;
    pkgs = nixpkgs.legacyPackages.x86_64-linux;
    llvmPkgs = pkgs.llvmPackages_15;
  in {
    devShells.x86_64-linux.default = pkgs.mkShell {
      packages = [
        llvmPkgs.clang

        llvmPkgs.libcxx
        llvmPkgs.libcxxabi
      ];
    };
  };
}
```
The same also occurs when using a shell with only `llvmPkgs.libcxxClang` and when trying to use `libcxx` inside a nix build.

I am pretty sure that this is caused by `lib/libc++.so` in `llvmPackages_15.libcxx` containing `INPUT(libc++.so.1 -l)` instead of `INPUT(libc++.so.1 -lc++abi)` (which it does for `llvmPackages_14.libcxx`). I don't know yet, what the root cause for that is or how to fix that.

### Notify maintainers
<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->
@rrbutani @Ericson2314 @sternenseemann @lovek323 @dtzWill @primeos

(So it shows up in the tracking issue https://github.com/NixOS/nixpkgs/issues/213033)

> Thanks for looking into this! Building from the PR-branch now no longer fails in the way I described and the example works for me again.
> 
> I have now found a new issue, tho: Importing `cxxlib.h` fails with:
> 
> ```
> /nix/store/qwnvng0cbyx0bijm654jpmpl0516hfhx-libcxxabi-15.0.7-dev/include/cxxabi.h:20:10: fatal error: '__cxxabi_config.h' file not found
> ```
> 
> For example, this `main.cpp` causes this error (using the same flake as above, just built from the PR branch).
> 
> ```c
> #include<cxxabi.h>
> int main() {}
> ```
> 
> I was able to work around this, by adding
> 
> ```shell
> install -m 644 ../../${pname}/include/__cxxabi_config.h "$dev/include"
> ```
> 
> to the end of the `postInstall` phase of `libcxxabi` locally, but I am not sure if that's the right approach here.

this change is missing from 15 https://github.com/nixos/nixpkgs/commit/6ba1b5b0b3e57a7def4e837f5599aa3ac68589a8#diff-10d812ae0081a7211c31f286a68130450af9ff7efa30fff60a337022816df01dR264

--- a/pkgs/development/compilers/llvm/15/libcxx/default.nix
+++ b/pkgs/development/compilers/llvm/15/libcxx/default.nix
@@ -1,5 +1,5 @@
 { lib, stdenv, llvm_meta
-, monorepoSrc, runCommand
+, monorepoSrc, runCommand, fetchpatch
 , cmake, ninja, python3, fixDarwinDylibNames, version
 , cxxabi ? if stdenv.hostPlatform.isFreeBSD then libcxxrt else libcxxabi
 , libcxxabi, libcxxrt, libunwind
@@ -47,6 +47,15 @@ stdenv.mkDerivation rec {
 
   patches = [
     ./gnu-install-dirs.patch
+    # See:
+    #   - https://reviews.llvm.org/D133566
+    #   - https://github.com/NixOS/nixpkgs/issues/214524#issuecomment-1429146432
+    # !!! Drop in LLVM 16+
+    (fetchpatch {
+      url = "https://github.com/llvm/llvm-project/commit/57c7bb3ec89565c68f858d316504668f9d214d59.patch";
+      hash = "sha256-AaM9A6tQ4YAw7uDqCIV4VaiUyLZv+unwcOqbakwW9/k=";
+      relative = "libcxx";
+    })
   ] ++ lib.optionals stdenv.hostPlatform.isMusl [
     ../../libcxx-0001-musl-hacks.patch
   ];

--- a/pkgs/test/default.nix
+++ b/pkgs/test/default.nix
@@ -59,7 +59,6 @@ with pkgs;
 
             # libcxxStdenv broken
             # fix in https://github.com/NixOS/nixpkgs/pull/216273
-            (filter (n: n != "llvmPackages_15"))
           ] ++ lib.optionals (stdenv.hostPlatform.isLinux && stdenv.hostPlatform.isAarch64) [
             # libcxx does not build for some reason on aarch64-linux
             (filter (n: n != "llvmPackages_7"))

